Adding docs for migrating from SDKv2 to the Framework (#461)
hashicorp/terraform-plugin-framework: website/docs/plugin/framework/migrating/attributes-blocks/validators-predefined.mdx

```go
func resourceExample() *schema.Resource {
	return &schema.Resource{
 		...
        Schema: map[string]*schema.Schema{
            "attribute_example": {
                ConflictsWith:  []string
                ExactlyOneOf:	[]string
                AtLeastOneOf:   []string
                RequiredWith:   []string
...
```
## Framework

In the Framework, you implement either type of validation by setting the `Validators` field on the `tfsdk.Attribute`
struct with types that satisfy the `tfsdk.AttributeValidator` interface. Validators that perform the same checks as the
predefined validators in SDKv2 are
[available for the Framework](https://github.com/hashicorp/terraform-plugin-framework-validators). If the predefined
validators do not meet your needs, you must define
[custom validators](/plugin/framework/migrating/validators-custom).

```go
func (d *resourceTypeExample) GetSchema(context.Context) (tfsdk.Schema, diag.Diagnostics) {
	return tfsdk.Schema{
        ...
		Attributes: map[string]tfsdk.Attribute{
            "attribute_example": {
                Validators: []tfsdk.AttributeValidator{

> 					schemavalidator.ConflictsWith(),

Reviewer: Over in https://github.com/hashicorp/terraform-plugin-framework-validators/pull/60 we'll also add support for "config" validators for this type of functionality that may be a little bit more intuitive for provider developers since you declare them outside a specific attribute. I think showing the 1:1 is nice here, but it might also be worth calling those out in the future since I think they'll be preferable.

Author: I've called out the "config" validators and added a generic example for usage with resources.

Reviewer: You rock, thank you

---
page_title: 'Attribute predefined validators: Migrating from SDKv2 to the Framework'
description: >-
  Validations check required syntax, types, and acceptable values.
  Migrate the predefined ConflictsWith, ExactlyOneOf, AtLeastOneOf and RequiredWith validators to the Framework.
---

# Validators - Predefined

Attribute validators ensure that attributes do or do not contain specific values. You can use predefined validators for
many use cases, or implement custom validators. Refer to [Schemas - Validators](/plugin/framework/schemas#validators) in
the Framework documentation for details. Refer to the
[Attributes - Custom Validators](/plugin/framework/migrating/validators-custom) page in this guide to learn how to
implement custom validators.

This page explains how to migrate a predefined validator from SDKv2 to the Framework.

## SDKv2

In SDKv2, the `ConflictsWith`, `ExactlyOneOf`, `AtLeastOneOf`, and `RequiredWith` fields on an attribute's
`schema.Schema` struct perform predefined validations on the list of attributes set for these fields.

```go
func resourceExample() *schema.Resource {
    return &schema.Resource{
        /* ... */

>         Schema: map[string]*schema.Schema
>         }{

Reviewer 2: ```suggestion
        Schema: map[string]*schema.Schema{
```

?

            "attribute_example": {
                ConflictsWith:  []string
                ExactlyOneOf:   []string
                AtLeastOneOf:   []string

>                 RequiredWith:   []string

Reviewer 2: Not familiar with this syntax - should these lines end with `,`?

Author: Fixed

            /* ... */
```
## Framework

In the Framework, you implement either type of validation by setting the `Validators` field on the `tfsdk.Attribute`
struct with types that satisfy the `tfsdk.AttributeValidator` interface. Validators that perform the same checks as the
predefined validators in SDKv2 are
[available for the Framework](https://github.com/hashicorp/terraform-plugin-framework-validators). If the predefined
validators do not meet your needs, you must define
[custom validators](/plugin/framework/migrating/validators-custom).

```go
func (d *resourceTypeExample) GetSchema(context.Context) (tfsdk.Schema, diag.Diagnostics) {
    return tfsdk.Schema{
        /* ... */
        Attributes: map[string]tfsdk.Attribute{
            "attribute_example": {
                Validators: []tfsdk.AttributeValidator{
                    schemavalidator.ConflictsWith(),
                    /* ... */
```

Configuration validators can also be defined for

> [providers](https://github.com/hashicorp/terraform-plugin-framework-validators/tree/main/providervalidator),
> [resources](https://github.com/hashicorp/terraform-plugin-framework-validators/tree/main/resourcevalidator) and
> [data sources](https://github.com/hashicorp/terraform-plugin-framework-validators/tree/main/datasourcevalidator) by

Reviewer: Nit: Prefer pkg.go.dev links over GitHub links

```suggestion
[providers](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework-validators/providervalidator),
[resources](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework-validators/resourcevalidator) and
[data sources](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework-validators/datasourcevalidator) by
```

> implementing `ProviderWithConfigValidators`, `ResourceWithConfigValidators` and `DataSourceWithConfigValidators`
> interfaces, respectively.

Reviewer 2: ```suggestion
implementing `ProviderWithConfigValidators`, `ResourceWithConfigValidators`, and `DataSourceWithConfigValidators`
interfaces, respectively.
```

Oxford comma best comma.

```go
func (r *resourceExample) ConfigValidators(ctx context.Context) []resource.ConfigValidator {
    return []resource.ConfigValidator{
        resourcevalidator.Conflicting(
            /* ... */
        ),
        /* ... */
```

## Migration Notes

Remember the following details when migrating from SDKv2 to the Framework.

- In SDKv2, `ValidateDiagFunc` is a field on `schema.Schema` that you can use to define validation functions. In SDKv2,
there are also built-in validations. For example, `ConflictsWith` is a field on the `schema.Schema` struct in SDKv2. In
the Framework, `Validators` is a field on each `tfsdk.Attribute` struct.

> - Validators replicating the behavior of `ConflictsWith`, `ExactlyOneOf`, `AtLeastOneOf` and `RequiredWith` in SDKv2 are

Reviewer 2: ```suggestion
- Validators replicating the behavior of `ConflictsWith`, `ExactlyOneOf`, `AtLeastOneOf`, and `RequiredWith` in SDKv2 are
```

available for the Framework:

>   - [ConflictsWith](https://github.com/hashicorp/terraform-plugin-framework-validators/blob/v0.4.0/schemavalidator/conflicts_with.go)
>   - [ExactlyOneOf](https://github.com/hashicorp/terraform-plugin-framework-validators/blob/v0.4.0/schemavalidator/exactly_one_of.go)
>   - [AtLeastOneOf](https://github.com/hashicorp/terraform-plugin-framework-validators/blob/v0.4.0/schemavalidator/at_least_one_of.go)
>   - [AlsoRequires](https://github.com/hashicorp/terraform-plugin-framework-validators/blob/v0.4.0/schemavalidator/also_requires.go)

Reviewer: Similarly:

```suggestion
  - [ConflictsWith](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework-validators/schemavalidator#ConflictsWith)
  - [ExactlyOneOf](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework-validators/schemavalidator#ExactlyOneOf)
  - [AtLeastOneOf](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework-validators/schemavalidator#AtLeastOneOf)
  - [AlsoRequires](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework-validators/schemavalidator#AlsoRequires)
```
